File: lib/setup/write.php
<?php
/*
Copyright (©) 2003-2014 Teus Benschop.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


error_reporting (E_ALL);


$messages = array ();


// Test write access to the exports folder.
$folder = realpath ("../exports");
if (!is_writable ($folder)) {
  $messages [] = "Bibledit-Web needs write access to folder $folder for exporting Bibles to.";
}


// Test write access to the git folder.
$folder = realpath ("../git");
if (!is_writable ($folder)) {
  $messages [] = "Bibledit-Web needs write access to folder $folder to store git repositories for collaboration.";
}


// Test write access to the revisions folder.
$folder = realpath ("../revisions");
if (!is_writable ($folder)) {
  $messages [] = "Bibledit-Web needs write access to folder $folder to store the daily changes.";
}


// Test write access to the dyncss folder.
$folder = realpath ("../dyncss");
if (!is_writable ($folder)) {
  $messages [] = "Bibledit-Web needs write access to folder $folder to store the dynamically created stylesheets.";
}


if (count ($messages) == 0) {
  include ("../filter/url.php");
  Filter_Url::redirect ("step.php?referer=write");
  die;
}


?>
<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<title>Bibledit-Web Installation</title>
<link rel="stylesheet" href="stylesheet.css" type="text/css" />
</head>
<body>
<h1><img src="../assets/bibledit.png"> <a href="http://bibledit.org/">Bibledit-Web</a></h1>
<p>Bibledit-Web tried to gain write access to certain locations on the web server, but did not succeed.</p>
<p>More information is below.</p>
<ul>
<?php
foreach ($messages as $message) {
  echo "<li>$message</li>\n";
}
?>
</ul>
<p>Your hosting provider will be able to advise what to do with regard to the problems mentioned.</p>
<p><a href="write.php" class="button button-large">Retry</a></p>
</body>
</html>

File: android/assets/external/manage/write.cpp
/*
Copyright (©) 2003-2015 Teus Benschop.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


#include <manage/write.h>
#include <assets/view.h>
#include <assets/page.h>
#include <filter/roles.h>
#include <filter/url.h>
#include <filter/string.h>
#include <webserver/request.h>
#include <database/books.h>
#include <locale/translate.h>


string manage_write_url ()
{
  return "manage/write";
}


bool manage_write_acl (void * webserver_request)
{
  return Filter_Roles::access_control (webserver_request, Filter_Roles::manager ());
}


string manage_write (void * webserver_request)
{
  Webserver_Request * request = (Webserver_Request *) webserver_request;

  string page;

  page = Assets_Page::header (translate ("Read/write"), webserver_request);

  Assets_View view = Assets_View ();

  string user = request->query["user"];
  view.set_variable ("user", user);
  
  string bible = request->query["bible"];
  view.set_variable ("bible", bible);

  int book = convert_to_int (request->query["book"]);

  bool access = request->database_users ()->hasAccess2Bible (user, bible);

  // Toggle write access to Bible.
  if (request->query.count ("toggle")) {
    if (access) {
      bool readonly = request->database_users ()->hasReadOnlyAccess2Book (user, bible, book);
      request->database_users ()->setReadOnlyAccess2Book (user, bible, book, !readonly);
    }
  }

  // Read or write access to display.
  vector <string> tbody;
  vector <int> books = request->database_bibles ()->getBooks (bible);
  
  for (size_t i = 0; i < books.size (); i++) {
    int book = books[i];
    int row = (i % 4);
    if (row == 0) tbody.push_back ("<tr>");
    if (row > 0) tbody.push_back ("<td> | </td>");
    tbody.push_back ("<td>");
    tbody.push_back (Database_Books::getEnglishFromId (book));
    tbody.push_back ("</td>");
    tbody.push_back ("<td>");
    tbody.push_back ("<a href=\"?user=" + user + "&bible=" + bible + "&book=" + convert_to_string (book) + "&toggle=\">");
    if (request->database_users ()->hasReadOnlyAccess2Book (user, bible, book)) {
      tbody.push_back (translate ("no"));
    } else {
      tbody.push_back (translate ("yes"));
    }
    tbody.push_back ("</a>");
    tbody.push_back ("</td>");
    if (row == 3) tbody.push_back ("</tr>");
  }
  if (books.empty ()) {
    tbody.push_back ("<tr>");
    tbody.push_back ("<td>");
    tbody.push_back (translate ("No books"));
    tbody.push_back ("</td>");
    tbody.push_back ("<td>");
  }
  
  view.set_variable ("tbody", filter_string_implode (tbody, "\n"));
  
  page += view.render ("manage", "write");

  page += Assets_Page::footer ();

  return page;
}
